readme.md
# Remote Logging

Call `initLogger` to override `console.log`, `console.warn`, `console.error` and `console.info` functions which will be routed to the given remote server.

```shell
npm install @ionic/remote-log
```

```typescript
import { initLogger } from '@ionic/remote-log';

initLogger('192.168.1.130:9000');
```



changelog.md
# Changelog

## 0.0.2
- Fix typescript error TS4111

index.js
"use strict";var _log,_warn,_error,_info,_deviceIdentifier,_serverUrl,__spreadArray=this&&this.__spreadArray||function(o,e,r){if(r||2===arguments.length)for(var n,i=0,t=e.length;i<t;i++)!n&&i in e||((n=n||Array.prototype.slice.call(e,0,i))[i]=e[i]);return o.concat(n||Array.prototype.slice.call(e))},_pending=(exports.__esModule=!0,exports.initLogger=void 0);function initLogger(o){var e;_log=window.console.log,_warn=window.console.error,_error=window.console.error,_info=window.console.info,_serverUrl=o,window.console.log=consoleLog,window.console.warn=consoleWarn,window.console.error=consoleError,window.console.info=consoleInfo,post("/devices",{id:getDeviceIdentifier(),agent:window.navigator.userAgent,title:window.document.title}),setInterval(function(){document.location.href!=e&&(e=document.location.href,_log("Url changed to ".concat(e)))},1e3)}function getDeviceIdentifier(){var o,e;return _deviceIdentifier||(o=localStorage.IonicLoggerDeviceId,e=parseInt(o),null!=o&&!isNaN(e)||(e=Math.floor(999999999*Math.random()),localStorage.IonicLoggerDeviceId=e),_deviceIdentifier=e.toString()),_deviceIdentifier}function write(o,e,r){var e=Array.prototype.slice.call(e),n=o;e.forEach(function(o){""!=n&&(n+=" "),n+="object"==typeof o?JSON.stringify(o):o}),_pending||(_pending=[],setTimeout(function(){post("/log",_pending),_pending=void 0},500)),_pending.push({id:getDeviceIdentifier(),message:n,level:r})}function getStack(){var o=(new Error).stack,o=null==o?void 0:o.split("\n");return null!=o&&o.splice(0,4),o&&0!=o.length?o[0].substr(7,o[0].length-7):""}function post(o,e){if(e)try{fetch("http://".concat(_serverUrl).concat(o),{method:"post",headers:{"Content-Type":"application/json"},body:JSON.stringify(e)})}catch(o){}}function consoleLog(o){for(var e=[],r=1;r<arguments.length;r++)e[r-1]=arguments[r];_log.apply(void 0,__spreadArray([o],e,!1)),write(o,e,"log")}function consoleWarn(o){for(var e=[],r=1;r<arguments.length;r++)e[r-1]=arguments[r];_warn.apply(void 0,__spreadArray([o],e,!1)),write(o,e,"warn")}function consoleError(o){for(var e=[],r=1;r<arguments.length;r++)e[r-1]=arguments[r];_error.apply(void 0,__spreadArray([o],e,!1)),write(o,e,"error")}function consoleInfo(o){for(var e=[],r=1;r<arguments.length;r++)e[r-1]=arguments[r];_info.apply(void 0,__spreadArray([o],e,!1)),write(o,e,"info")}exports.initLogger=initLogger;

index.ts
let _log: Function;
let _warn: Function;
let _error: Function;
let _info: Function;
let _deviceIdentifier: string;
let _serverUrl: string;
let _pending: Array<LogItem> | undefined = undefined;

interface LogItem {
    id: string;
    message: string;
    level: string;
}

/**
 * Initialize logging will override window.console and send to the serverUrl
 * @param  {string} serverUrl The servername and port number of the remote server (eg 192.168.1.1:9000)
 */
export function initLogger(serverUrl: string) {
    _log = window.console.log;
    _warn = window.console.error;
    _error = window.console.error;
    _info = window.console.info;
    _serverUrl = serverUrl;
    window.console.log = consoleLog;
    window.console.warn = consoleWarn;
    window.console.error = consoleError;
    window.console.info = consoleInfo;

    let lastUrl: string;
    post('/devices', {
        id: getDeviceIdentifier(),
        agent: window.navigator.userAgent,
        title: window.document.title,
    });

    // Report urls
    setInterval(() => {
        if (document.location.href != lastUrl) {
            lastUrl = document.location.href;
            _log(`Url changed to ${lastUrl}`);
        }
    }, 1000);
}

function getDeviceIdentifier(): string {
    if (_deviceIdentifier) {
        return _deviceIdentifier;
    }
    const tmp = localStorage['IonicLoggerDeviceId'];
    let id: number = parseInt(tmp);
    if (tmp == null || isNaN(id)) {
        // Create a random device identifier
        id = Math.floor(Math.random() * 999999999);
        localStorage['IonicLoggerDeviceId'] = id;
    }
    _deviceIdentifier = id.toString();
    return _deviceIdentifier;
}

function write(message: any, _arguments: any, level: string) {
    const args = Array.prototype.slice.call(_arguments);
    let msg = message;
    args.forEach((element) => {
        if (msg != '') {
            msg += ' ';
        }
        if (typeof element == 'object') {
            msg += JSON.stringify(element);
        } else {
            msg += element;
        }
    });
    // Commenting out for now. Stack is hard as it may be in the source map
    //const stack = this.getStack();

    if (!_pending) {
        _pending = [];
        setTimeout(() => {
            // Push pending log entries. We wait around for 500ms to see how much accumulates
            post('/log', _pending);
            _pending = undefined;
        }, 500);        
    }
    _pending.push({ id: getDeviceIdentifier(), message: msg, level: level }); // this.getStack() });
}

function getStack(): string {
    const stack = new Error().stack;
    const lines = stack?.split('\n');
    lines?.splice(0, 4);
    if (!lines || lines.length == 0) {
        return '';
    }
    return lines[0].substr(7, lines[0].length - 7); // This returns just the top of the stack
}

function post(url: string, data: any) {
    if (!data) {
        return;
    }

    try {
        fetch(`http://${_serverUrl}${url}`, {
            method: 'post',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(data),
        });
    } catch {
        // Logging should not cause failures
    }
}

function consoleLog(message: any, ...args: any) {
    _log(message, ...args);
    write(message, args, 'log');
}

function consoleWarn(message: any, ...args: any) {
    _warn(message, ...args);
    write(message, args, 'warn');
}

function consoleError(message: any, ...args: any) {
    _error(message, ...args);
    write(message, args, 'error');
}

function consoleInfo(message: any, ...args: any) {
    _info(message, ...args);
    write(message, args, 'info');
}